src/serializers/shared.rs

```diff
@@ -195,7 +197,20 @@ impl CombinedSerializer {
         }
 
         let type_: Bound<'_, PyString> = schema.get_as_req(type_key)?;
-        Self::find_serializer(type_.to_str()?, schema, config, definitions)
+        let type_ = type_.to_str()?;
+
+        // if we have a SchemaValidator on the type already, use it
+        // however, we don't want to use a prebuilt validator for dataclasses if we have a generic_origin
+        // because __pydantic_serializer__ is cached on the unparametrized dataclass
+        if matches!(type_, "model" | "typed-dict")
+            || matches!(type_, "dataclass") && !schema.contains(intern!(py, "generic_origin"))?
```

Comment:
I somewhat wonder if this check would be better made inside of `PrebuiltSerializer::build`